src/views/ApartmentApp/components/StudentApplication/index.js

```diff
@@ -0,0 +1,884 @@
+  const [applicants, setApplicants] = useState([]);
+  const [preferredHalls, setPreferredHalls] = useState([]); // Properties 'HallName' and 'HallRank' must be capitalized to match the backend
+
+  const [applicationCardsOpen, setApplicationCardsOpen] = useState(false);
+  const [newEditorProfile, setNewEditorProfile] = useState(null); // Stores the StudentProfileInfo of the new editor before the user confirms the change
+  const [changeEditorDialogOpen, setChangeEditorDialogOpen] = useState(false);
+  const [submitDialogOpen, setSubmitDialogOpen] = useState(false); // Use this for submitting app (later feature)
+  const [snackbarOpen, setSnackbarOpen] = useState(false);
+  const [snackbarText, setSnackbarText] = useState('');
+  const [snackbarSeverity, setSnackbarSeverity] = useState('');
+  const [saveButtonAlertTimeout, setSaveButtonAlertTimeout] = useState(null);
+
+  /**
+   * Loads the user's saved apartment application, if one exists
+   */
+  const loadSavedApplication = useCallback(async () => {
+    // TODO: Implement this once save/load of application data has been implemented in the backend
+    setLoading(true);
+    // Check if the current user is on an application. Returns the application ID number if found
+    let newApplicationID = null;
+    try {
+      newApplicationID = await housing.getCurrentApplicationID();
+      //! DEBUG
+      console.log('Retrieved Application ID: ' + newApplicationID);
+      if (newApplicationID === null || newApplicationID === -1) {
+        // Intentionally trigger the 'catch'
+        throw new Error("Invalid value of 'newApplicationID' = " + newApplicationID);
+      }
+      setApplicationID(newApplicationID);
+      let applicationDetails = await housing.getApartmentApplication(newApplicationID);
+      if (applicationDetails) {
+        if (applicationDetails.DateSubmitted) {
+          setDateSubmitted(applicationDetails.DateSubmitted);
+        }
+        if (applicationDetails.DateModified) {
+          setDateModified(applicationDetails.DateModified);
+        }
+        if (applicationDetails.Username) {
+          setEditorUsername(applicationDetails.Username);
+        }
+        if (applicationDetails.Applicants) {
+          applicationDetails.Applicants.forEach(async (applicantInfo) => {
+            const newApplicantProfile = await user.getProfileInfo(applicantInfo.Username);
+            setApplicants((prevApplicants) =>
+              prevApplicants.concat({
+                Profile: newApplicantProfile,
+                OffCampusProgram: applicantInfo.OffCampusProgram,
+              }),
+            );
+          });
+        }
+      }
+    } catch {
+      // No existing application was found in the database,
+      // or an error occurred while attempting to load the application
+      setApplicationID(-1);
+      if (!editorUsername) {
+        if (
+          applicants.every((applicant) => applicant.Profile.AD_Username !== userProfile.AD_Username)
```

Comment:
This is really minor but I think we could use `!applicants.some(...)` instead of `applicants.every(...)`. This should have a minor benefit to performance since it will finish earlier in the case that a matching applicant is found, and I think it's also (arguably) clearer becuase it literally means 'if there isn't some applicant that matches this username'

Reply:
I realize now that since you're doing a negative comparison, the shortcutting behavior is already achieved, so then it comes down to a syntax/semantics opinion, and that isn't as important. What you have already works well.

Reply:
Typically this part only gets run the first time the page is loaded (or at least that is the intent), so the applicants array is expected to be empty or one element...

This makes me realized I could actually just compare `applicants[0].Profile.AD_Username` since the application editor is always the first element of the array by design

Reply:
It technically should not need to check this at all, since it is meant to happen only with the editor has not been set at the page load, 
but regardless of whether I put the `setEditorUsername` before or after that `setApplicants` line, there was always a race condition of it somehow triggering nearly simultaneously and causing it to enter the current user into the array twice on page load. 
So the applicants array check is to prevent that race condition

Reply:
The race condition you're describing I believe is the normal behavior of useState. Updates to state are batched and run altogether according to some complex React-internal logic. This prevents excessive rerenders, because otherwise every set[WhateverStateVariableName]() call would cause an independent rerender.

Usually, if you have state that is derived from other state, you want a useEffect callback that will which the state that is derived from and update the derived state appropriately. 

I'm not sure I fully understand this usecase, so I can't say for sure whether that's what you need here, but hopefully that's helpful.

Reply:
I believe I have fixed this in a more recent push, but the best solution is just check if the array is empty,

It is basically saying (through the combination of all the nested if/else) that if there is not an existing application, then this must be a new application, so make sure the current user gets added into the array. 

Two answer your other comment,
The other problem (which has been a big problem in other parts of the Apartapp stuff) is that the useEffect will only do a shallow comparison if the "derived state" is an array of custom objects, so it actually often does not handle the updates correctly (either triggering at the wrong time or failing to trigger when an object within the array should be effected, etc.)

I think in this use case, the simplest solutions is the say
```
if (!applicants.length) {
  // insert the current user 
}
```
Since the goal is simply to insert the current user as an applicant iff the array hasn't been filled in yet.

Reply:
Now that I think about it, I guess this could be split up in various way, such as doing it outside the `if (!editorUsername)` or even before the call of the load endpoint, since the entire applicant array is replaced by whatever is loaded from the backend (if an application is found), otherwise it should just initialize to only the current user.

Or, it may be possible to initialize that via the `= useState(......)` that initializes the state variables